src/language/patterns.md

```diff
@@ -0,0 +1,334 @@
+---
+title: Patterns
+description: Summary of patterns in Dart.
+---
+
+{{site.alert.version-note}}
+  Patterns require a [language version][] of at least 3.0.
+{{site.alert.end}}
+
+Patterns are a syntactic category in the Dart language, like statements and expressions.
+They represent the form of a value, not an actual value.
+
+This page describes:
+- What patterns do (match and destructure)
+- Where patterns are allowed in Dart code
+- Common use cases for patterns
+
+To learn about the different types of patterns, visit the [pattern type reference][types].
+
+## Matching
+
+A pattern is always tested against a value to determine if the value has the form
+you expect, or in other words, if the value _matches_ the pattern. 
+
+What constitutes a match depends on the [pattern type][types].
+For example, matching a constant pattern means the value is equivalent to the pattern:
+
+```dart
+switch (number) {
+  case 1: ...        // Constant pattern 1 matches if number == 1
+}
+```
+
+Many patterns make use of subpatterns (sometimes called outer and inner patterns,
+respectively). Patterns match recursively on their subpatterns.
+For example, the individual fields of any [collection-type][] pattern could be 
+[variable patterns][variable] or [constant patterns][constant]:
+
+```dart
+switch (obj) {       // List pattern [a,b] matches obj first if obj
+  case [a, b]: ...   // is a list wth two fields, then if its  
+}                    // fields match the constant subpatterns a and b  
+```
+
+Patterns can also _partially_ match, by using a [wildcard pattern][] as a place holder,
+or a [rest element][] in the case of list patterns.
+
+If you need to ensure `null` values aren't silently treated as match failures,
+use the [null-check pattern][] while matching.
+
+## Destructuring
+
+When an object and pattern match, the pattern can then access the object's data 
+and extract it in parts. In other words, the pattern _destructures_ the object:
+
+```dart
+var numList = [1, 2, 3];
+var [a, b, c] = numList;   // List pattern [a, b, c] destructures the list object
+print(a + b + c);          // numList into individual new parts. Prints 6.
+```
+
+You can use the [null-check pattern][] to check for `null` while destructuring. 
```

Comment:
Well, this gets to my comment below.  You can use the null check pattern to destructure a value in a switch, but not in a destructuring assignment (which I think is more what this section is about).

Reply:
Or maybe you don't indent this to just be about destructuring assignment?  But then it seems somewhat redundant with the previous section?  I guess I'm not clear on how these two sections fit together.

Reply:
The `(int?, int?)` syntax there is a record type not a pattern.  The `p?` syntax meaning "nullable type" as a type and "check for null" is unfortunately very confusing - we just never found anything better.

Reply:
Oh, ok. I wonder if this warrants explaining somewhere, though I'm not sure where that explanation would go.